### PR TITLE
Fixed #357.

### LINKED ISSUE #357: Conversion with default values
msgpack::object provides `as<>()` and `convert(T)` functions to extract values. In order to handle fields that are potentially nil, newer msgpack revisions also provide support for `boost::optional`, so I can write code like this:
```
auto myint = object.as<boost::optional<int>>();
if (myint) {
    // Perform code based on myint existing.
}
```

There is still an unnecessary verbosity when dealing with default values:
```
std::map<std::string, whatever_t> mymap;
if (!object.is_nil()) {
    object.convert(mymap);
}
int myint = DEFAULT_NUMBER;
if (!object.is_nil()) {
    object.convert(myint);
}
// Or a slightly more verbose example:
std::vector<int> list = foo->number_list();
if (!object.is_nil()) {
    auto list = object.convert(list);
    foo->set_number_list(list);
}
```

I can see two potential accessor method patterns that can help simplify this code:
```
// get_or() (named after boost::optional::get_or()) provides a fallback in case the object is nil.
auto mystring = object.get_or(std::string());
auto mymap = object.get_or(std::map<std::string, whatever_t>());

// For implicit int/uint conversions, get_or() can be dangerous if the caller forgets to also provide the template argument manually:
uint64_t counter1 = object.get_or(0);
// The above resolves to int as template argument, which eliminates much of the uint64_t range.

// The other approach resembles convert(), and lets the user do the initialization of default values explicitly:
uint64_t counter2 = 0;
object.convert_if_not_nil(counter2);

// Having a bool return value for convert_if_not_nil() also helps to reduce the number of lines:
std::vector<int> list = foo->number_list();
if (object.convert_if_not_nil(list)) {
    foo->set_number_list(list);
}
```

All of the above have in common that `object` is only used once, which can sometimes help prevent creation of a temporary msgpack::object variable and just continue from a msgpack::object function result instead. Which can make the code even more compact (while retaining readability) compared to the baseline.

While the `get_or()` suffers from possible template mismatches and can't be used for `if (...) foo->set(...);` style clauses, `convert_if_not_nil()` handles both well and seems like an improvement without drawbacks. Here is an implementation (note that it will still throw if the object exists but contains an unexpected type):

```
template <typename T>
inline bool object::convert_if_not_nil(T& v) const
{
    if (is_nil()) {
        return false;
    }
    v = as<T>();
    return true;
}
```

How would you feel about this method being added to msgpack::object?

### PR DESCRIPTION
**Pull request by redboltz**
Added a conversion member function to msgpack::object.
If msgpack::object is nil then returns false else returns true and sets a value.

### CODE REVIEW

**Comment by redboltz** [comment]
@jpetso, could you review this PR?

File: include/msgpack/object.hpp [hunk 1]
```diff
@@ -511,6 +511,15 @@ inline T* object::convert(T* v) const
     return v;
 }
 
+template <typename T>
+inline bool object::convert_if_not_nil(T& v) const {
```
Comment (jpetso) [inline include/msgpack/object.hpp]: Style suggestion: In other methods, you put the brace on a separate line. Maybe here too.

**Comment by jpetso** [comment]
Awesome! r+

File: include/msgpack/object.hpp (hunk 1 above)
Comment (redboltz) [inline include/msgpack/object.hpp]: Fixed.

**Comment by redboltz** [comment]
Thank you for reviewing. Merged.

**Pull request merged**